lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -219,20 +355,70 @@ static LogicalResult extractConcatToConcatExtract(ExtractOp op,
   return success();
 }
 
-LogicalResult ExtractOp::canonicalize(ExtractOp op, PatternRewriter &rewriter) {
-  // Narrow Mux
-  // extract(mux(c,a,b)) -> mux(c,extract(a),extract(b))
-  if (auto mux = dyn_cast_or_null<MuxOp>(op.input().getDefiningOp())) {
-    if (mux->hasOneUse()) {
-      auto newT = rewriter.createOrFold<ExtractOp>(
-          mux.getLoc(), op.getType(), mux.trueValue(), op.lowBit());
-      auto newF = rewriter.createOrFold<ExtractOp>(
-          mux.getLoc(), op.getType(), mux.falseValue(), op.lowBit());
-      rewriter.replaceOpWithNewOp<MuxOp>(op, mux.cond(), newT, newF);
-      return success();
-    }
+// Pattern matches on extract(f(a, b)), and transforms f(extract(a), extract(b))
+// for some known f. This is performed by analyzing all usage sites of the
+// result of f(a, b). When this transformation returns true, it mutates all the
+// usage sites of outerExtractOp to reference the newly created narrowed
+// operation.
+static bool narrowExtractWidth(ExtractOp outerExtractOp,
+                               PatternRewriter rewriter) {
+  auto *innerArg = outerExtractOp.input().getDefiningOp();
+
+  if (!innerArg) {
+    return false;
   }
 
+  // In calls to narrowOperationWidth below, innerOp is guranteed to have at
+  // least one use (ie: this extract operation). So we don't need to handle
+  // innerOp with no uses.
+
+  return llvm::TypeSwitch<Operation *, bool>(innerArg)
+      // The unreferenced leading bits of Add, Sub and Mul can be stripped of,
+      // but not the trailing bits. The trailing bits is used to compute the
+      // results of arithmetic operations.
+      .Case<AddOp, MulOp>([&](auto innerOp) {
+        return narrowOperationWidth(outerExtractOp, innerOp, innerOp.inputs(),
+                                    /* narrowTrailingBits= */ false, rewriter);
+      })
+      .Case<SubOp>([&](auto innerOp) {
```

Comment:
Not sure if you have a preference, but for the single type cases I'm pretty sure you could also do:

 `.Case([&](SubOp innerOp) ...`